### ISSUE TITLE
"for x in arr" breaks auto-vectorization, unlike semantically-equivalent indexing

### ISSUE THREAD

**Issue by itamarst** [numba/numba#9210]
Numba is 0.58. llvmlite is 0.41.0. Ubuntu 22.04 on i7-12700K.

Consider the following script:

```python
from timeit import timeit
from numba import njit
import numpy as np

ARR = np.ones((1_00, 1_00), dtype=np.uint64)

@njit
def variant1(arr):
    total = 0
    for value in arr.flatten():
        total += value
    return total

@njit
def variant2(arr):
    total = 0
    arr = arr.flatten()
    for i in range(len(arr)):
        total += arr[i]
    return total

assert variant1(ARR) == variant2(ARR)

print("variant1", timeit(lambda: variant1(ARR)))
print("variant2", timeit(lambda: variant2(ARR)))
```

If I run this with auto-vectorization disabled, they take the same amount of time:

```
$ NUMBA_LOOP_VECTORIZE=0 python benchmark2.py 
variant1 4.528854038973805
variant2 4.5288080560276285
```

With auto-vectorization enabled, however, the second version is much faster:

```
$ python benchmark2.py 
variant1 5.6276919129886664
variant2 2.4880486209876835
```

## Discussion

It's possible this isn't fixable. But if Numba knows the array is 1-dimensional, it could generate LLVM IR for `variant1` equivalent to `variant2` (instead of whatever it's generating now), to help LLVM optimize it? They do seem semantically identical to me so long as the indexing variable is anonymoys, though I might be missing something.

**Comment by sklam** [comment]
Right, the two variants are semantically identical. It is possible to implement such rewrite in Numba, however, the current Numba infrastructure is not designed to be friendly for such transformations.  The Numba devs are working on new intermediate representation  (https://github.com/numba/numba-rvsdg) so that high-level rewrites are easier to implement. 

Another problem for doing rewrites is the combinatorial problem. Rewrites rules can conflict with each other or require a specific ordering to work. This leads to the team investigating in Equality-Saturation and Equivalent Graphs (https://egraphs-good.github.io/)

**Comment by sklam** [comment]
From triage meeting:

The two variants used to have similar performance until a patch for iterating over ND-array was merged. 
@stuartarchibald may remember the PR number.

**Comment by itamarst** [comment]
I did some further experiments, where I tried to find idioms that would help the compiler optimize away bounds checking. Based on these experiments, I think the preferred idiom is actually:

```python
i = 0
while i < len(arr):
    total += arr[i]
```

Or, and this is silly looking but it really does both optimize away bounds checking _and_ enable auto-vectorization:

```python
for i, _ in enumerate(arr):
    total += arr[i]
```
